NPE at cgeo.geocaching.utils.CollectionStream.forEach called by LogCacheActivity.resetValues

Error marked as "new with beta 2", currently 18 occurrences, Android 9 only, two different phone models

```
java.lang.RuntimeException: 
  at android.app.ActivityThread.performLaunchActivity (ActivityThread.java:2944)
  at android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:3079)
  at android.app.servertransaction.LaunchActivityItem.execute (LaunchActivityItem.java:78)
  at android.app.servertransaction.TransactionExecutor.executeCallbacks (TransactionExecutor.java:108)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:68)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1836)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:193)
  at android.app.ActivityThread.main (ActivityThread.java:6702)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:493)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:911)
Caused by: java.lang.NullPointerException: 
  at cgeo.geocaching.utils.CollectionStream.forEach (CollectionStream.java:174)
  at cgeo.geocaching.log.LogCacheActivity.resetValues (LogCacheActivity.java:405)
  at cgeo.geocaching.log.LogCacheActivity.onCreate (LogCacheActivity.java:278)
  at android.app.Activity.performCreate (Activity.java:7166)
  at android.app.Activity.performCreate (Activity.java:7157)
  at android.app.Instrumentation.callActivityOnCreate (Instrumentation.java:1271)
  at android.app.ActivityThread.performLaunchActivity (ActivityThread.java:2924)
```

--- a/main/src/cgeo/geocaching/log/LogCacheActivity.java
+++ b/main/src/cgeo/geocaching/log/LogCacheActivity.java
@@ -402,7 +402,8 @@ public class LogCacheActivity extends AbstractLoggingActivity {
         final EditText logPasswordView = LogCacheActivity.this.findViewById(R.id.log_password);
         logPasswordView.setText(StringUtils.EMPTY);
 
-        CollectionStream.of(trackables).forEach(tl -> initializeTrackableAction(tl, null));
+        //force copy due to #9101
+        CollectionStream.of(trackables, true).forEach(tl -> initializeTrackableAction(tl, null));
     }
 
     private void clearLog() {

--- a/main/src/cgeo/geocaching/utils/CollectionStream.java
+++ b/main/src/cgeo/geocaching/utils/CollectionStream.java
@@ -3,6 +3,8 @@ package cgeo.geocaching.utils;
 import cgeo.geocaching.utils.functions.Action1;
 import cgeo.geocaching.utils.functions.Func1;
 
+import androidx.annotation.NonNull;
+
 import java.lang.reflect.Array;
 import java.util.Arrays;
 import java.util.Collection;
@@ -35,7 +37,7 @@ import java.util.stream.Collectors;
 @SuppressWarnings("unchecked")
 public class CollectionStream<T> {
 
-    private final Collection<?> originalCollection;
+    private final Collection<Object> originalCollection;
     //LinkedList can be used most efficiently for 'map' and 'filter' methods
     private LinkedList<Object> collection;
 
@@ -43,18 +45,32 @@ public class CollectionStream<T> {
      * creates CollectionStream with a Collection as its source
      */
     public static <TT> CollectionStream<TT> of(final Collection<TT> coll) {
-        return new CollectionStream<>(coll);
+        return of(coll, false);
+    }
+
+    /**
+     * creates CollectionStream with a Collection as its source
+     * if forceCopy is true, then given coll is copied even if only read operations are performed on it.
+     */
+    public static <TT> CollectionStream<TT> of(final Collection<TT> coll, final boolean forceCopy) {
+        return new CollectionStream<>((Collection<Object>) coll, forceCopy);
     }
 
     /**
      * creates CollectionStream with an array as its source
+     *
      */
     public static <TT> CollectionStream<TT> of(final TT[] coll) {
-        return new CollectionStream<>(Arrays.asList(coll));
+        return new CollectionStream<>(Arrays.asList(coll), false);
     }
 
-    private CollectionStream(final Collection<?> coll) {
-        this.originalCollection = coll == null ? Collections.emptyList() : coll;
+    private CollectionStream(final Collection<Object> coll, final boolean forceCopy) {
+        if (forceCopy) {
+            this.originalCollection = Collections.emptyList(); // do not store link to original collection
+            this.collection = new LinkedList<>(coll == null ? Collections.emptyList() : coll);
+        } else {
+            this.originalCollection = coll == null ? Collections.emptyList() : coll;
+        }
     }
 
     /**
@@ -102,7 +118,6 @@ public class CollectionStream<T> {
         }
         return this;
     }
-
     /**
      * mimics {@link java.util.stream.Collectors#joining()}
      */
@@ -125,7 +140,6 @@ public class CollectionStream<T> {
         }
         return sb.toString();
     }
-
     /**
      * mimics {@link Collectors#toList()}
      */
@@ -176,10 +190,15 @@ public class CollectionStream<T> {
         }
     }
 
+    @NonNull
     private Collection<Object> getCollectionForRead() {
-        return this.collection == null ? (Collection<Object>) this.originalCollection : this.collection;
+        if (this.collection != null) {
+            return this.collection;
+        }
+        return this.originalCollection == null ? Collections.emptyList() : this.originalCollection;
     }
 
+    @NonNull
     private LinkedList<Object> getCollectionForWrite() {
         if (this.collection == null) {
             this.collection = new LinkedList<>(this.originalCollection);

--- a/tests/src/cgeo/geocaching/utils/CollectionStreamTest.java
+++ b/tests/src/cgeo/geocaching/utils/CollectionStreamTest.java
@@ -1,5 +1,6 @@
 package cgeo.geocaching.utils;
 
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashMap;
@@ -28,6 +29,16 @@ public class CollectionStreamTest {
         assertThat(CollectionStream.of((Collection<Object>) null).toJoinedString(",")).isEqualTo("");
         assertThat(CollectionStream.of(new Integer[]{1, 2}).map(null).filter(null).toJoinedString(",")).isEqualTo("1,2");
         assertThat(CollectionStream.of(new Integer[]{1, 2}).toJoinedString(null)).isEqualTo("12");
+    }
+
+    @Test
+    public void testReadOnModifiedCollections() {
+        final List<String> list = new ArrayList<>(Arrays.asList(new String[]{"one", "two"}));
+        final CollectionStream<String> csWithCopy = CollectionStream.of(list, true);
+        final CollectionStream<String> csWithoutCopy = CollectionStream.of(list, false);
+        list.add("three");
+        assertThat(csWithCopy.toJoinedString(",")).isEqualTo("one,two");
+        assertThat(csWithoutCopy.toJoinedString(",")).isEqualTo("one,two,three");
     }
 
     @Test

